wasm/src/decoder/util.rs

```diff
@@ -0,0 +1,50 @@
+use super::common;
+use crate::error::Error;
+use crate::error::Result;
+use leb128::read;
+
+pub(crate) fn decode_vec<T, R, F>(reader: &mut R, func: F) -> Result<Vec<T>>
+where
+    R: std::io::Read,
+    F: Fn(&mut R) -> Result<T>,
+{
+    let length = read::unsigned(reader)? as usize;
+    if length > u32::MAX as usize {
+        return Err(Error::ArrayTooLarge);
+    }
+
+    let mut v = Vec::with_capacity(length);
+
+    for x in v.iter_mut().take(length) {
+        *x = func(reader)?;
+    }
+
+    Ok(v)
+}
+
+pub fn decode_kind<R: std::io::Read>(reader: &mut R) -> Result<common::ValueKind> {
+    let mut byte: [u8; 1] = [0; 1];
+    reader.read_exact(&mut byte)?;
+
+    match byte[0].try_into() {
+        Ok(v) => Ok(v),
+        Err(_) => Err(Error::InvalidValueKind),
+    }
+}
+
+pub fn decode_u32<R: crate::Reader>(reader: &mut R) -> Result<u32> {
+    let s = reader.seek(std::io::SeekFrom::Current(0))?;
+    let length = read::unsigned(reader)?;
+    // This is so wacky to do. Replacing it with a better system should be done at some point
+    let end = reader.seek(std::io::SeekFrom::Current(0))?;
```

Comment:
I wouldn't expect this to even move the cursor, wouldn't you want to save the last position?

Reply:
well we would need to determine how many bytes were read. and by subtracting the begin position from the end we should get the difference. Eg how many bytes were read than expected